File: Python Implementation/core/expressions.py
class Expression:
    def accept(self):
        raise NotImplementedError("Accept method not implemented in base Expression class")

class NumberExpression(Expression):
    def __init__(self, value):
        self.value = value

    def accept(self, visitor):
        return visitor.visit_number(self)

class VariableExpression(Expression):
    def __init__(self, name):
        self.name = name

    def accept(self, visitor):
        return visitor.visit_variable(self)

class UnaryExpression(Expression):
    def __init__(self, operator, operand):
        self.operator = operator
        self.operand = operand

    def accept(self, visitor):
        return visitor.visit_unary(self)

class PostfixExpression(Expression):
    def __init__(self, operand, operator):
        self.operand = operand
        self.operator = operator

    def accept(self, visitor):
        return visitor.visit_postfix(self)

class BinaryExpression(Expression):
    def __init__(self, left, operator, right):
        self.left = left
        self.operator = operator
        self.right = right

    def accept(self, visitor):
        return visitor.visit_binary(self)

class CallExpression(Expression):
    def __init__(self, identifier, arguments):
        self.identifier = identifier
        self.arguments = arguments

    def accept(self, visitor):
        return visitor.visit_call(self)

File: src/cpp/core/expressions.cpp
#include "expressions.h"
#include "evaluator.h"
#include <memory>

NumberExpression::NumberExpression(double value) : value(value) {}
VariableExpression::VariableExpression(FixedString16 name) : name(name) {}
UnaryExpression::UnaryExpression(SyntaxToken op, std::unique_ptr<Expression> operand) : op(op), operand(std::move(operand)) {}
PostfixExpression::PostfixExpression(std::unique_ptr<Expression> operand, SyntaxToken op) : operand(std::move(operand)), op(op) {}
BinaryExpression::BinaryExpression(std::unique_ptr<Expression> left, SyntaxToken op, std::unique_ptr<Expression> right) : left(std::move(left)), op(op), right(std::move(right)) {}
CallExpression::CallExpression(FixedString16 functionName, std::vector<std::unique_ptr<Expression>> arguments) : functionName(functionName), arguments(std::move(arguments)) {}

double NumberExpression::accept(Evaluator& visitor) {
    return visitor.visit(*this);
}

double VariableExpression::accept(Evaluator& visitor) {
    return visitor.visit(*this);
}

double UnaryExpression::accept(Evaluator& visitor) {
    return visitor.visit(*this);
}

double PostfixExpression::accept(Evaluator& visitor) {
    return visitor.visit(*this);
}

double BinaryExpression::accept(Evaluator& visitor) {
    return visitor.visit(*this);
}

double CallExpression::accept(Evaluator& visitor) {
    return visitor.visit(*this);
}